cmake: kconfig: Minor command-line assignment improvements

57300 opened the pull request:
Summary:

1. Preserve boolean (=n) assignments from command-line.
2. Discard invalid assignments from command-line.
3. Avoid writing a new `.config` when nothing changes.

57300 committed:
cmake: kconfig: Keep symbol names sorted in EXTRA_KCONFIG_OPTIONS

This makes checksum calculation over Kconfig fragments more consistent,
which prevents writing a new `.config` when nothing really changes.

To explain this, consider the following sequence:

1. west build . -DCONFIG_XXXX=y -DCONFIG_YYYY=y
2. west build . -DCONFIG_YYYY=y
3. west build . -DCONFIG_XXXX=y

At (1), we set new values for XXXX and YYYY, so the `.config` changes.

At (2), we set a value for YYYY, but it's the same value as before, so
the `.config` doesn't get overwritten.

At (3), we set a value for XXXX, but it's the same value as before, so
the `.config` shouldn't get overwritten... but it does. What happened?

The reason is that the generated `extra_kconfig_options.conf` fragment,
which is included in the checksum calculation, was being populated using
two sets of CMake cache variables:

- past assignments, prefixed with `CLI_${KCONFIG_NAMESPACE}_`.
- new assignments, prefixed with just `${KCONFIG_NAMESPACE}_`.

Usually, past assignments would appear before new assignments, because
the default `${KCONFIG_NAMESPACE}` is CONFIG, which goes after CLI in
alphabetical order. As a result, the contents of EXTRA_KCONFIG_OPTIONS
at (1) and (2):

   CONFIG_XXXX=y
   CONFIG_YYYY=y

were not identical to its contents at (3):

   CONFIG_YYYY=y
   CONFIG_XXXX=y

resulting in a different checksum.

This is resolved by stripping out the CLI prefix first, then effectively
"mergesorting" the past and new assignments, before starting to populate
EXTRA_KCONFIG_OPTIONS.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

tejlmand reviewed (changes requested):
Thanks for discovering and providing a fix.

Before approving I would like you to take a look at adjusting `import_kconfig()` instead, as I believe making that function behaving identical to Kconfiglib is a safer approach.

Couple of references into kconfiglib:
https://github.com/zephyrproject-rtos/zephyr/blob/34e50037f0d94efce10770da5abcf48f7ba33906/scripts/kconfig/kconfiglib.py#L982-L983
https://github.com/zephyrproject-rtos/zephyr/blob/34e50037f0d94efce10770da5abcf48f7ba33906/scripts/kconfig/kconfiglib.py#L1266
https://github.com/zephyrproject-rtos/zephyr/blob/34e50037f0d94efce10770da5abcf48f7ba33906/scripts/kconfig/kconfiglib.py#L1325
https://github.com/zephyrproject-rtos/zephyr/blob/34e50037f0d94efce10770da5abcf48f7ba33906/scripts/kconfig/kconfiglib.py#L1338
https://github.com/zephyrproject-rtos/zephyr/blob/34e50037f0d94efce10770da5abcf48f7ba33906/scripts/kconfig/kconfiglib.py#L1347

57300 committed:
cmake: extensions: Improve import_kconfig() bool/tristate parsing

This function has been working fine for parsing Kconfig fragments as
processed by Kconfiglib, but it can be made more robust for parsing
fragments from other sources, including handwritten ones.

Previously, Kconfig fragment lines of the form `# CONFIG_FOO is not set`
were ignored entirely, while lines of the form `CONFIG_FOO=n` would set
the corresponding CMake variable or target property to a literal `n`.
However, Kconfiglib treats both equivalently - as assignments to `n` -
so `import_kconfig()` should too.

Due to the fact that `.config` files output by Kconfiglib always show
disabled options as `# CONFIG_FOO is not set` (which was being ignored),
existing CMake code expects `CONFIG_FOO` to be unset. To avoid breakage,
the variable/property will now be unset explicitly by `import_kconfig()`
when it encounters either representation of `CONFIG_FOO=n`.

Moreover, for bool and tristate symbols, Kconfiglib accepts assignments
like `CONFIG_FOO=yeah`, `CONFIG_FOO=maybe`, or `CONFIG_FOO=nope`, and it
transforms the value of `CONFIG_FOO` into `y`, `m`, or `n` respectively.
This effect is now replicated in CMake.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
cmake: kconfig: Preserve correct CLI assignments across re-runs

This tiny patch makes two improvements:

1. Preserve boolean (=n) assignments from command-line.

This fixes an issue where, if a symbol with `default y` were turned off
via command-line, e.g., `-DCONFIG_BOOT_BANNER=n`, a CMake re-run would
revert that symbol back to its default value.

To avoid this, the assignment should have been preserved in CMake cache
as `CLI_CONFIG_BOOT_BANNER:INTERNAL=n`. However, `kconfig.cmake` clears
unset variables from cache, and (=n) symbols become unset variables, so
an exception had to be made for them.

2. Discard invalid assignments from command-line.

Although `kconfig.cmake` takes care to discard assignments to symbols
which get unset by Kconfig, it wasn't handling the case where Kconfig
would keep the symbol but replace its value, making the CMake-cached
assignment invalid.

For example, this assignment:

   west build . -DCONFIG_PRINTK=n

could be invalidated if PRINTK were selected by, e.g., BOOT_BANNER,
producing this warning:

   PRINTK (...) was assigned the value 'n' but got the value 'y'. (...)

Still, the old value of (=n) was being cached. One way in which this was
evident was when setting an unrelated symbol in a separate invocation:

   west build . -DCONFIG_MAIN_STACK_SIZE=512

the same warning for PRINTK would show up again.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 commented on cmake/modules/kconfig.cmake:
```diff
@@ -379,7 +379,21 @@ import_kconfig(${KCONFIG_NAMESPACE} ${DOTCONFIG})
+    set(possible_values "n")
+  endif()
+
+  if(value IN_LIST possible_values)
```
I submitted an alternative solution inspired by your suggestion.

I agree that `import_kconfig()` could be made more robust, so I added another patch for that here. Thanks for the tips.

There's only one detail that I'm unsure about. In the case of either:
```
# CONFIG_FOO is not set
CONFIG_FOO=n
```
did you want the CMake variable `CONFIG_FOO` to be set to a literal `n`, rather than unset? Because that would be a breaking change.

tejlmand replied:
> I submitted an alternative solution inspired by your suggestion.

Great.

> There's only one detail that I'm unsure about. In the case of either:
> ```
> # CONFIG_FOO is not set
> CONFIG_FOO=n
> ```

The `CONFIG_FOO=n` version is not valid in the `.config` file.
If it is present, then kconfiglib will transform that line into `# CONFIG_FOO is not set`.

This means that the form `CONFIG_FOO=n` can only be present in `.config` if a user manually has modified the file, and in such case will be short lived as the next run of kconfiglib (which is triggered by CMake) will update the setting and hence the line will be `# CONFIG_FOO is not set` when `import_kconfig()` executes.

Also note, if a user manually modifies the `.config` then it's in principle the user's responsibility that the format is correct.

> did you want the CMake variable CONFIG_FOO to be set to a literal n, rather than unset? Because that would be a breaking change.

That is a valid point.
For consistency we should probably set it to `n` explicitly, and the way we generally use `CONFIG_<X>` in CMake it should work.
But as there could be occurrences where someone does a `if(DEFINED CONFIG_<X>)`, then keeping current behavior to unset is the safest approach, so let's stick to that with this PR.

I believe changing behavior is better done in a separate PR, where pros / cons can be discussed.

nordicjm replied:
Whilst that change would be nice, it's a bit "too little too late". If that change was made 5 years ago then it would have been fine, but to make that change _now_ would cause **major** breakage both in and out of tree, needing a lot of things to be reworked and I'm not in favour of that.

tejlmand replied:
> but to make that change now would cause major breakage both in and out of tree

Not sure I would consider it a major breakage.

If someone does `if(DEFINED CONFIG_FOO)` when they want to test if `FOO=y` or `if(NOT DEFINED CONFIG_FOO)` when the assume `FOO=n` they are doing it wrong.

If `FOO` is considered a boolean, then it ought to be checked as a bool, and not rely on a defined variable being equal to a true value.

But I acknowledge that to many developers then the difference between `if(CONFIG_FOO)` and `if(DEFINED CONFIG_FOO)` may go unnoticed as long as the CMake code behaves as they intend.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1552,40 +1552,62 @@ endfunction()
-    # The variable name match we just did included the '=' symbol. To just get the
-    # part on the RHS we use match group 1
-    set(CONF_VARIABLE_VALUE ${CMAKE_MATCH_1})
+    # If the provided value is n, then the corresponding CMake variable or
+    # target property will be unset.
+    if("${CONF_VARIABLE_VALUE}" STREQUAL "n")
+      if(DEFINED IMPORT_KCONFIG_TARGET)
+        set_property(TARGET ${IMPORT_KCONFIG_TARGET} PROPERTY "${CONF_VARIABLE_NAME}")
+      else()
+        unset("${CONF_VARIABLE_NAME}" PARENT_SCOPE)
+      endif()
```
I see that this part could benefit from config settings being internally represented by `n` instead of an `undef`.

I think having a prepare PR which changes the current behavior from `# CONFIG_<var> is not set` being undefined into being a `n` would make sense, and doing it as a prepare PR will allow us to run CI and check code and side-effect independent of other changes in this PR.

57300 replied:
> I see that this part could benefit from config settings being internally represented by `n` instead of an `undef`.

Sure. The way I wrote it, you could eventually remove this entire if-block and get the `n` behavior that you want.

tejlmand commented on cmake/modules/kconfig.cmake:
```diff
@@ -376,15 +376,32 @@ foreach (name ${cli_config_list})
+foreach (name ${cache_variable_names})
+  # Note: "${CLI_${name}}" is the verbatim value of ${name} from command-line,
+  # while "${temp_${name}}" is the same value processed by 'import_kconfig'.
+  if(((NOT DEFINED ${name}) AND (NOT DEFINED temp_${name})) OR
+     ((DEFINED ${name}) AND (DEFINED temp_${name}) AND (${name} STREQUAL temp_${name})))
```
I believe this code can be cleaner if we know that all settings (both those explicitly set, and those with `# CONFIG_FOO is not set`) results in a defined variable in CMake.

57300 replied:
Agreed.

tejlmand reviewed (commented):
Looking good, but before giving +1, I would like to take the opportunity to discuss if we should make the transition of `# CONFIG_<var> is not set` resulting in an undefined CMake variable, to instead be a defined variable with value `n`, as mentioned here https://github.com/zephyrproject-rtos/zephyr/pull/59915#discussion_r1281893032.

Such a change is better taken in separate PR, and I think would be nice to do before this PR, but such a change can of course also be done afterwards.
Having such change will also mean that we can differentiate between variables `not set` and those not even present in the `.config`, as setting not present will still be undefined.

57300 commented:
> Looking good, but before giving +1, I would like to take the opportunity to discuss if we should make the transition of `# CONFIG_<var> is not set` resulting in an undefined CMake variable, to instead be a defined variable with value `n`, as mentioned here [#59915 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/59915#discussion_r1281893032).

My take is that the most dangerous thing about this change is probably how it would affect things out of tree. That's where I see the most occurrences of this pattern: `if([NOT] DEFINED CONFIG_...)`, which you already pointed out as problematic.

> Having such change will also mean that we can differentiate between variables `not set` and those not even present in the `.config`, as setting not present will still be undefined.

Sounds interesting, but can you think of a good use case for this? I think in most cases, it's natural to simply infer that e.g., `CONFIG_LOG_RUNTIME_FILTERING` is undefined if it's guarded by `CONFIG_LOG=n`.

> doing it as a prepare PR will allow us to run CI and check code and side-effect independent of other changes in this PR.

I'm not sure if CI will necessarily catch the more subtle side-effects, like unused code being compiled in.

tejlmand reviewed (approved):
approved.

Nothing blocking in this PR, and if we want to align the CMake handling of `CONFIG_FOO=n` to that described by kconfiglib then that can be done later and code pointed out in this PR can be improved.
